src/redis/windowed_cache_impl.go

```diff
@@ -0,0 +1,263 @@
+	*pipeline = client.PipeAppend(*pipeline, nil, "EXPIRE", key, expirationSeconds)
+	*pipeline = client.PipeAppend(*pipeline, result, "GET", key)
+}
+
+func windowedSetNewTatPipelineAppend(client Client, pipeline *Pipeline, key string, newTat int64, expirationSeconds int64) {
+	*pipeline = client.PipeAppend(*pipeline, nil, "SET", key, newTat)
+	*pipeline = client.PipeAppend(*pipeline, nil, "EXPIRE", key, expirationSeconds)
+}
+
+func (this *windowedRateLimitCacheImpl) DoLimit(
+	ctx context.Context,
+	request *pb.RateLimitRequest,
+	limits []*config.RateLimit) []*pb.RateLimitResponse_DescriptorStatus {
+
+	logger.Debugf("starting windowed cache lookup")
+
+	// request.HitsAddend could be 0 (default value) if not specified by the caller in the Ratelimit request.
+	hitsAddend := max(1, request.HitsAddend)
+
+	// First build a list of all cache keys that we are actually going to hit. GenerateCacheKey()
+	// returns an empty string in the key if there is no limit so that we can keep the arrays
+	// all the same size.
+	assert.Assert(len(request.Descriptors) == len(limits))
+	cacheKeys := make([]limiter.CacheKey, len(request.Descriptors))
+	for i := 0; i < len(request.Descriptors); i++ {
+		cacheKeys[i] = this.cacheKeyGenerator.GenerateCacheKey(
+			request.Domain, request.Descriptors[i], limits[i], 0)
+
+		// Increase statistics for limits hit by their respective requests.
+		if limits[i] != nil {
+			limits[i].Stats.TotalHits.Add(uint64(hitsAddend))
+		}
+	}
+
+	// Get existing tat value for each cache keys
+	isOverLimitWithLocalCache := make([]bool, len(request.Descriptors))
+	tats := make([]int64, len(request.Descriptors))
+	var pipeline, perSecondPipeline Pipeline
+	for i, cacheKey := range cacheKeys {
+		if cacheKey.Key == "" {
+			continue
+		}
+
+		if this.localCache != nil {
+			// Get returns the value or not found error.
+			_, err := this.localCache.Get([]byte(cacheKey.Key))
+			if err == nil {
+				isOverLimitWithLocalCache[i] = true
+				logger.Debugf("cache key is over the limit: %s", cacheKey.Key)
+				continue
+			}
+		}
+
+		logger.Debugf("looking up tat for cache key: %s", cacheKey.Key)
+
+		expirationSeconds := utils.UnitToDivider(limits[i].Limit.Unit)
+
+		// Use the perSecondConn if it is not nil and the cacheKey represents a per second Limit.
+		if this.perSecondClient != nil && cacheKey.PerSecond {
+			if perSecondPipeline == nil {
+				perSecondPipeline = Pipeline{}
+			}
+			windowedPipelineAppend(this.perSecondClient, &perSecondPipeline, cacheKey.Key, &tats[i], expirationSeconds)
+		} else {
+			if pipeline == nil {
+				pipeline = Pipeline{}
+			}
+			windowedPipelineAppend(this.client, &pipeline, cacheKey.Key, &tats[i], expirationSeconds)
+		}
+	}
+
+	if pipeline != nil {
+		checkError(this.client.PipeDo(pipeline))
+		pipeline = nil
+	}
+	if perSecondPipeline != nil {
+		checkError(this.perSecondClient.PipeDo(perSecondPipeline))
+		perSecondPipeline = nil
+	}
+
+	// Rate limit GCRA logic
```

Comment:
This code needs more comments per above. Maybe an overview comment will help me.

Also, a bunch of non-trivial code in this file is shared with the fixed impl. Is it possible to share more code via a shared utility or mix-in?

Reply:
Hi, can you please give me examples in ratelimit about the shared utility and mix-in?
Thanks